File: Day10-19/10.py
from collections import defaultdict


def process_bots(cmds):
    bots, output, edges = defaultdict(list), defaultdict(list), {}
    for cmd in map(lambda x: x.split(), cmds):
        if cmd[0] == 'value':
            bots[int(cmd[-1])].append(int(cmd[1]))
        else:
            l, h = map(int, (cmd[6], cmd[-1]))
            edges[int(cmd[1])] = (l, cmd[5], h, cmd[-2])
    while bots:
        try:
            bot, inv = next((k, v) for k, v in bots.iteritems() if len(v) == 2)
        except StopIteration:
            break
        bots[bot] = []
        l, l_io, h, h_io = edges[bot]
        if 61 in inv and 17 in inv:
            print "Part 1: %s" % bot
        (bots if l_io == 'bot' else output)[l].append(min(inv))
        (bots if h_io == 'bot' else output)[h].append(max(inv))
    return {k: v[0] for k, v in output.items()}


if __name__ == '__main__':
    with open('10-test.txt', 'r') as f:
        process_bots(f.readlines())

    with open('10.txt', 'r') as f:
        o = process_bots(f.readlines())
        print("Part 2: %s" % (o[0] * o[1] * o[2]))

File: Day10-19/10.swift
//
//  10.swift
//  
//
//  Created by Benjamin Congdon on 12/12/16.
//
//

import Foundation

struct Edge {
    var lo: Int
    var hi: Int
    var isBotLo: Bool
    var isBotHi: Bool
}

func process_bots(commands: [String]) -> [Int] {
    var bots: [Int:[Int]] = [:]
    for i in (0...500){
        bots[i] = []
    }
    var output: [Int] = [Int](count: 100, repeatedValue: 0)
    var edges: [Int:Edge] = [:]
    for cmd in commands.map({$0.componentsSeparatedByString(" ")}) {
        if(cmd[0] == "value") {
            let idx = Int(cmd.last!)!
            if(bots[idx] != nil) {
                bots[idx]!.append(Int(cmd[1])!)
            }
            else{
                bots[idx] = [Int(cmd[1])!]
            }
        }
        else {
            let lo = Int(cmd[6])!
            let hi = Int(cmd.last!)!
            edges[Int(cmd[1])!] = Edge(lo: lo, hi: hi,
                isBotLo: cmd[5] == "bot", isBotHi: cmd[10] == "bot")
        }
    }
    while(bots.count > 0) {
        var bot = -1
        var inv = [Int]()
        for (k, v) in bots {
            if v.count == 2 {
                bot = k
                inv = v
                break
            }
        }
        if bot < 0 {
            break
        }
        bots[bot] = []
        if inv.contains(61) && inv.contains(17) {
            print("Part 1: " + String(bot))
        }
        let e = edges[bot]!
        if e.isBotLo {
            bots[e.lo]!.append(inv.minElement()!)
        }
        else{
            output[e.lo] = inv.minElement()!
        }
        if e.isBotHi {
            bots[e.hi]!.append(inv.maxElement()!)
        }
        else {
            output[e.hi] = inv.maxElement()!
        }
    }
    return output
}

let text: String
do {
    text = try String(contentsOfFile: "10.txt")    
}
catch {
    text = ""
}

let o = process_bots(text.componentsSeparatedByString("\n"))
print("Part 2: " + String(o[0] * o[1] * o[2]))